README.md
# Apple Health Clinical Records

Enables reading clinical records from Apple health.

|             | iOS   |
|-------------|-------|
| **Support** | 12.0+ |


This plugin supports:

- Checking if Apple health is available using the `checkIfHealthDataAvailable` method.
- handling permissions to access clinical records using the `hasAuthorization`, `requestAuthorization` methods.
- reding clinical records using the `getData` method.

## Data types

| **Data Type** | **FHIR resource**   | **Comments**       |
|---------------|---------------------|--------------------|
| condition     | Condition           |                    |
| allergy       | AllergyIntolerance  |                    |
| coverage      |                     | Required iOS 14.0+ |
| immunization  | Immunization        |                    |
| labResult     | Observation         |                    |
| medication    | MedicationOrder     |                    |
| procedure     | Procedure           |                    |
| vitalSign     | Observation         |                    |


## Setup

First, add `apple_health_clinical_records` as a [dependency in your pubspec.yaml file](https://flutter.dev/using-packages/).

### iOS

Step 1: Append the Info.plist with the following entry

```
<key>NSHealthClinicalHealthRecordsShareUsageDescription</key>
<string>Your clinical records usage description</string>
```

Step 2: Enable "HealthKit" inside the "Capabilities" tab.

Step 3: Enable "Clinical Health Records" capability for "HealthKit".

![Clinical Health Records in HealthKit](./resources/clinical_records.png)

## Usage

See the example app for detailed examples of how to use the Clinical records API.

The Clinical records plugin is used via the `AppleHealthClinicalRecords` class using the different methods for handling permissions and getting 
records from apple health.

```dart

// Create an instance of AppleHealthClinicalRecords
AppleHealthClinicalRecords clinicalRecords = AppleHealthClinicalRecords();

// Check if apple health is available
bool available =
    await clinicalRecords.checkIfHealthDataAvailable() ?? false;

// define the types to get
var types = [
  ClinicalType.allergy
];

// check if permission is given for defined types
final futures = types.map((type) => clinicalRecords.hasAuthorization(type));

final authorizations = await Future.wait(futures);

if (authorizations.every((element) => true)) {
  
  getClinicalData(types);
  
} else {
  
  // request authorizations
  bool authorized = await clinicalRecords.requestAuthorization(types) ?? false;
  
  if (authorized) {
    getClinicalData(types);
  }
  
}

void getClinicalData(List<ClinicalType> types) async {
  final records = types.map((type ) => clinicalRecords.getData(type));

  final data = await Future.wait(records);
}

```

### Clincal record

A clinical record contains list of FHIR resources, of listed data types (see above).

see [FHIR resources list](http://hl7.org/fhir/resourcelist.html)

### References

Inspired by [Health plugin](https://pub.dev/packages/health).

http://hl7.org/fhir/resourcelist.html

https://developer.apple.com/documentation/healthkit/samples/accessing_health_records

https://developer.apple.com/documentation/healthkit/samples/accessing_a_user_s_clinical_records



CHANGELOG.md
## 0.0.3

* Request read persmission to multiple types at once
* Check authorization status of a clinical type
* Fix issue only returning random single record
* Updated readme instructions

## 0.0.2

* Defined supported clinical record types
* Android platform check

## 0.0.1

* TODO: Describe initial release.


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility in the flutter_test package. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:apple_health_clinical_records_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(const MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) =>
            widget is Text && widget.data!.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


test/apple_health_clinical_records_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:apple_health_clinical_records/apple_health_clinical_records.dart';
import 'package:apple_health_clinical_records/apple_health_clinical_records_platform_interface.dart';
import 'package:apple_health_clinical_records/apple_health_clinical_records_method_channel.dart';
import 'package:plugin_platform_interface/plugin_platform_interface.dart';

class MockAppleHealthClinicalRecordsPlatform
    with MockPlatformInterfaceMixin
    implements AppleHealthClinicalRecordsPlatform {
  @override
  Future<bool?> checkIfHealthDataAvailable() => Future.value(true);

  @override
  Future<bool?> requestAuthorization(List<String> sampleType) {
    throw UnimplementedError();
  }

  @override
  Future<bool?> hasAuthorization(String type) {
    throw UnimplementedError();
  }

  @override
  Future getData(String sampleType) {
    throw UnimplementedError();
  }
}

void main() {
  final AppleHealthClinicalRecordsPlatform initialPlatform =
      AppleHealthClinicalRecordsPlatform.instance;

  test('$MethodChannelAppleHealthClinicalRecords is the default instance', () {
    expect(initialPlatform,
        isInstanceOf<MethodChannelAppleHealthClinicalRecords>());
  });

  test('getPlatformVersion', () async {
    AppleHealthClinicalRecords appleHealthClinicalRecordsPlugin =
        AppleHealthClinicalRecords();
    MockAppleHealthClinicalRecordsPlatform fakePlatform =
        MockAppleHealthClinicalRecordsPlatform();
    AppleHealthClinicalRecordsPlatform.instance = fakePlatform;

    expect(await appleHealthClinicalRecordsPlugin.checkIfHealthDataAvailable(),
        '42');
  });
}


test/apple_health_clinical_records_method_channel_test.dart
import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:apple_health_clinical_records/apple_health_clinical_records_method_channel.dart';

void main() {
  MethodChannelAppleHealthClinicalRecords platform =
      MethodChannelAppleHealthClinicalRecords();
  const MethodChannel channel = MethodChannel('apple_health_clinical_records');

  TestWidgetsFlutterBinding.ensureInitialized();

  setUp(() {
    channel.setMockMethodCallHandler((MethodCall methodCall) async {
      return '42';
    });
  });

  tearDown(() {
    channel.setMockMethodCallHandler(null);
  });

  test('getPlatformVersion', () async {
    expect(await platform.checkIfHealthDataAvailable(), '42');
  });
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'dart:async';

import 'package:flutter/services.dart';
import 'package:apple_health_clinical_records/apple_health_clinical_records.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatefulWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  final _appleHealthClinicalRecordsPlugin = AppleHealthClinicalRecords();

  @override
  void initState() {
    super.initState();
    initPlatformState();
  }

  Future<void> initPlatformState() async {
    try {
      // First we check if Apple Health is available or not
      final available = await _appleHealthClinicalRecordsPlugin
              .checkIfHealthDataAvailable() ??
          false;

      debugPrint(available ? 'available' : 'not available');

      if (available) {
        // Then check the authorization status of required clinical types

        final futures = ClinicalType.values.map(
            (type) => _appleHealthClinicalRecordsPlugin.hasAuthorization(type));

        final statuses = await Future.wait(futures);

        var authStatus = false;

        if (statuses.any((element) => element != true)) {
          // Some or more clinical types are not authorized

          var types = [ClinicalType.allergy];

          authStatus = await _appleHealthClinicalRecordsPlugin
                  .requestAuthorization(ClinicalType.values) ??
              false;
        }

        debugPrint(authStatus.toString());

        if (authStatus) {
          final records = ClinicalType.values
              .map((type) => _appleHealthClinicalRecordsPlugin.getData(type));

          final data = await Future.wait(records);

          debugPrint(data.toString());
        }
      }
    } on PlatformException catch (err) {
      debugPrint(err.toString());
    }

    // If the widget was removed from the tree while the asynchronous platform
    // message was in flight, we want to discard the reply rather than calling
    // setState to update our non-existent appearance.
    if (!mounted) return;
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Plugin example app'),
        ),
        body: Center(
          child: Text('Running on:'),
        ),
      ),
    );
  }
}


lib/apple_health_clinical_records.dart
import 'dart:io';

import 'apple_health_clinical_records_platform_interface.dart';

/// Types of supported clinical records
enum ClinicalType {
  allergy,
  condition,
  coverage,
  immunization,
  labResult,
  medication,
  procedure,
  vitalSign
}

/// Extension on [ClinicalType] to get string representation
extension ClinicalTypeX on ClinicalType {
  String get name {
    switch (this) {
      case ClinicalType.allergy:
        return 'allergy';
      case ClinicalType.condition:
        return 'condition';
      case ClinicalType.coverage:
        return 'coverage';
      case ClinicalType.immunization:
        return 'immunization';
      case ClinicalType.labResult:
        return 'labResult';
      case ClinicalType.medication:
        return 'medication';
      case ClinicalType.procedure:
        return 'procedure';
      case ClinicalType.vitalSign:
        return 'vitalSign';
    }
  }
}

/// Class to be instantiated to use the plugin methods
class AppleHealthClinicalRecords {
  /// Call to instance method [checkIfHealthDataAvailable]
  Future<bool?> checkIfHealthDataAvailable() {
    /// Android platform is not supported
    if (Platform.isAndroid) {
      return Future.value(false);
    }

    return AppleHealthClinicalRecordsPlatform.instance
        .checkIfHealthDataAvailable();
  }

  /// Call to instance method [requestAuthorization]
  Future<bool?> requestAuthorization(List<ClinicalType> types) {
    List<String> typesToRead = [];

    /// Converting [types] to their string representation
    for (var type in types) {
      typesToRead.add(type.name);
    }

    return AppleHealthClinicalRecordsPlatform.instance
        .requestAuthorization(typesToRead);
  }

  /// Call to instance method [hasAuthorization]
  Future<bool?> hasAuthorization(ClinicalType type) {
    return AppleHealthClinicalRecordsPlatform.instance
        .hasAuthorization(type.name);
  }

  /// Call to instatnce method [getData]
  Future<dynamic> getData(ClinicalType sampleType) {
    return AppleHealthClinicalRecordsPlatform.instance.getData(sampleType.name);
  }
}


lib/apple_health_clinical_records_method_channel.dart
import 'package:flutter/foundation.dart';
import 'package:flutter/services.dart';

import 'apple_health_clinical_records_platform_interface.dart';

/// An implementation of [AppleHealthClinicalRecordsPlatform] that uses method channels.
class MethodChannelAppleHealthClinicalRecords
    extends AppleHealthClinicalRecordsPlatform {
  /// The method channel used to interact with the native platform.
  @visibleForTesting
  final methodChannel = const MethodChannel('apple_health_clinical_records');

  /// implementation of [checkIfHealthDataAvailable]
  @override
  Future<bool?> checkIfHealthDataAvailable() async {
    final status =
        await methodChannel.invokeMethod<dynamic>('checkIfHealthDataAvailable');
    return status;
  }

  /// implementation of [hasAuthorization]
  @override
  Future<bool?> hasAuthorization(String type) async {
    final status = await methodChannel
        .invokeMethod<dynamic>('hasAuthorization', {'type': type});
    return status;
  }

  /// implementation of [requestAuthorization]
  @override
  Future<bool?> requestAuthorization(List<String> types) async {
    final status = await methodChannel
        .invokeMethod<dynamic>('requestAuthorization', {'types': types});
    return status;
  }

  /// implementation of [getData]
  @override
  Future<dynamic> getData(String sampleType) async {
    final data = await methodChannel
        .invokeMethod<dynamic>('getData', {'sampleType': sampleType});
    return data;
  }
}


lib/apple_health_clinical_records_platform_interface.dart
import 'package:plugin_platform_interface/plugin_platform_interface.dart';

import 'apple_health_clinical_records_method_channel.dart';

/// A representation of plugin methods.
abstract class AppleHealthClinicalRecordsPlatform extends PlatformInterface {
  /// Constructs a AppleHealthClinicalRecordsPlatform.
  AppleHealthClinicalRecordsPlatform() : super(token: _token);

  static final Object _token = Object();

  static AppleHealthClinicalRecordsPlatform _instance =
      MethodChannelAppleHealthClinicalRecords();

  /// The default instance of [AppleHealthClinicalRecordsPlatform] to use.
  ///
  /// Defaults to [MethodChannelAppleHealthClinicalRecords].
  static AppleHealthClinicalRecordsPlatform get instance => _instance;

  /// Platform-specific implementations should set this with their own
  /// platform-specific class that extends [AppleHealthClinicalRecordsPlatform] when
  /// they register themselves.
  static set instance(AppleHealthClinicalRecordsPlatform instance) {
    PlatformInterface.verifyToken(instance, _token);
    _instance = instance;
  }

  /// Performs the check if apple health is available on device
  Future<bool?> checkIfHealthDataAvailable() {
    throw UnimplementedError(
        'checkIfHealthDataAvailable() has not been implemented.');
  }

  /// Request read only permission to [types]
  Future<bool?> requestAuthorization(List<String> types) {
    throw UnimplementedError(
        'requestAuthorization() has not been implemented.');
  }

  /// Check if [type] already has read permission
  Future<bool?> hasAuthorization(String type) {
    throw UnimplementedError('hasAuthorization() has not been implemented.');
  }

  /// Get clinical records of [sampleType]
  Future<dynamic> getData(String sampleType) {
    throw UnimplementedError('getData() has not been implemented.');
  }
}


ios/Classes/SwiftAppleHealthClinicalRecordsPlugin.swift
import Flutter
import UIKit
import HealthKit

struct PluginError: Error {
    let message: String
}

@available(iOS 12.0, *)
enum Section {
    case healthRecords

    var displayName: String {
        switch self {
        case .healthRecords:
            return "Health Records"
        }
    }

    var types: [HKSampleType] {
        switch self {
        case .healthRecords:
            return [
                HKObjectType.clinicalType(forIdentifier: .allergyRecord)!,
                HKObjectType.clinicalType(forIdentifier: .vitalSignRecord)!,
                HKObjectType.clinicalType(forIdentifier: .conditionRecord)!,
                HKObjectType.clinicalType(forIdentifier: .immunizationRecord)!,
                HKObjectType.clinicalType(forIdentifier: .labResultRecord)!,
                HKObjectType.clinicalType(forIdentifier: .medicationRecord)!,
                HKObjectType.clinicalType(forIdentifier: .procedureRecord)!
            ]

        }
    }
}

@available(iOS 12.0, *)
public class SwiftAppleHealthClinicalRecordsPlugin: NSObject, FlutterPlugin {

    let healthStore = HKHealthStore()

    var sampleTypes: Set<HKSampleType> {
        return Set(Section.healthRecords.types)
    }

    var sampleTypesDict: [String: HKSampleType] = [:]

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "apple_health_clinical_records", binaryMessenger: registrar.messenger())
        let instance = SwiftAppleHealthClinicalRecordsPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {

        initializeTypes()

        /// Handle checkIfHealthDataAvailable
        if (call.method.elementsEqual("checkIfHealthDataAvailable")){
            checkIfHealthDataAvailable(call: call, result: result)
        }

        else if (call.method.elementsEqual("hasAuthorization")){
            try! hasAuthorization(call: call, result: result)
        }
        /// Handle requestAuthorization
        else if (call.method.elementsEqual("requestAuthorization")){
            try! requestAuthorization(call: call, result: result)
        }

        /// Handle getData
        else if (call.method.elementsEqual("getData")){
           try! getData(call: call, result: result)
        }

    }

    func checkIfHealthDataAvailable(call: FlutterMethodCall, result: @escaping FlutterResult) {
        result(HKHealthStore.isHealthDataAvailable())
    }


    func hasAuthorization(call: FlutterMethodCall, result: @escaping FlutterResult) throws {

        let arguments = call.arguments as? NSDictionary


        guard let type = arguments?["type"] as? String

        else {
            throw PluginError(message: "Invalid Arguments!")
        }



        let status = healthStore.authorizationStatus(for: sampleTypesDict[type]!)

        if status == HKAuthorizationStatus.sharingAuthorized {
            result(true)
        } else {
            result(false)
        }


    }


    func requestAuthorization(call: FlutterMethodCall, result: @escaping FlutterResult) throws {

        let arguments = call.arguments as? NSDictionary


        guard let types = arguments?["types"] as? Array<String>

        else {
            throw PluginError(message: "Invalid Arguments!")
        }


        var typesToRead = Set<HKSampleType>()

        for (_, key) in types.enumerated() {
            guard let dataType = sampleTypesDict[key]

            else {
                throw PluginError(message: "Invalid Arguments!")
            }

            typesToRead.insert(dataType)
        }


        healthStore.requestAuthorization(toShare: Set(), read: typesToRead) { (success, error) in
            DispatchQueue.main.async {
                result(success)
            }
        }


    }


    func getData(call: FlutterMethodCall, result: @escaping FlutterResult) throws {
        let arguments = call.arguments as? NSDictionary

        guard let sampleTypeKey = arguments?["sampleType"] as? String

        else {
            throw PluginError(message: "Invalid Arguments!")
        }


        queryForSamples(sampleType: sampleTypesDict[sampleTypeKey]!, result: result)

    }


    /// Use HKSampleQuery to query the HealthKit store for samples by type.
    func queryForSamples(sampleType: HKSampleType, result: @escaping FlutterResult) {
        let sortDescriptors = [NSSortDescriptor(key: HKSampleSortIdentifierStartDate, ascending: false)]

        let today = Date()
        let dateFrom = Calendar.current.date(byAdding: .year, value: -3, to: today)


        let predicate = HKQuery.predicateForSamples(withStart: dateFrom, end: today, options: .strictEndDate)

        let query = HKSampleQuery(sampleType: sampleType, predicate: predicate, limit: 100, sortDescriptors: sortDescriptors) {(_, samplesOrNil, error) in
            DispatchQueue.main.async {
                guard let samples = samplesOrNil else {
                    return
                }

                let records = (samples as Any) as! [HKClinicalRecord]

                var dictionaries = Array<NSDictionary>();

                for record in records {
                    guard let fhirRecord = record.fhirResource else {
                        return
                    }

                    do {
                        let jsonDictionary = try JSONSerialization.jsonObject(with: fhirRecord.data, options: [])

                        dictionaries.append(jsonDictionary as! NSDictionary)
                    }
                    catch let error {
                        print("*** An error occurred while parsing the FHIR data: \(error.localizedDescription) ***")
                    }

                }


                result(dictionaries)
            }
        }

        healthStore.execute(query)
    }

    func initializeTypes() {


        sampleTypesDict["allergy"] = HKObjectType.clinicalType(forIdentifier: .allergyRecord)!

        sampleTypesDict["condition"] = HKObjectType.clinicalType(forIdentifier: .conditionRecord)!

        if #available(iOS 14.0, *) {
            sampleTypesDict["coverage"] = HKObjectType.clinicalType(forIdentifier: .coverageRecord)!
        } else {
            // Fallback on earlier versions
        }

        sampleTypesDict["immunization"] = HKObjectType.clinicalType(forIdentifier: .immunizationRecord)!

        sampleTypesDict["labResult"] = HKObjectType.clinicalType(forIdentifier: .labResultRecord)!

        sampleTypesDict["medication"] = HKObjectType.clinicalType(forIdentifier: .medicationRecord)!

        sampleTypesDict["procedure"] = HKObjectType.clinicalType(forIdentifier: .procedureRecord)!

        sampleTypesDict["vitalSign"] = HKObjectType.clinicalType(forIdentifier: .vitalSignRecord)!

    }


}
